wiki/Writing-tests-with-good-behavioural-coverage.md

```diff
@@ -0,0 +1,1265 @@
+    assertThat(checkNumber(Int.MIN_VALUE)).isEqualTo("Negative")
+}
+```
+
+## 5. Covering All Branches, Paths, and Conditions
+
+Testing all branches, paths, and conditions within your code is essential to ensure that every possible execution path is verified. This approach helps in identifying edge cases and logic errors that could otherwise go unnoticed. Effective testing covers all possible scenarios, including combinations of conditions and branching logic.
+
+Let's see the function to evaluate a user's access level based on their age and membership status.
+
+```
+fun evaluateAccess(age: Int, isMember: Boolean): String {
+    var result: String
+
+    if (age >= 18 && isMember) {
+        result = "Access granted"
+    } else if (age >= 18 && !isMember) {
+        result = "Membership required"
+    } else {
+        result = "Access denied"
+    }
+
+    return result
+}
+```
+
+The different scenarios and the expected outcomes are,
+
+| Scenario         | Description                           | Expected Outcome           |
+|------------------|---------------------------------------|----------------------------|
+| Adult Member     | `age >= 18` and `isMember = true`     | Returns "Access granted"   |
+| Adult Non-Member | `age >= 18` and `isMember = false`    | Returns "Membership required" |
+| Minor Member     | `age < 18` and `isMember = true`      | Returns "Access denied"    |
+| Minor Non-Member | `age < 18` and `isMember = false`     | Returns "Access denied"    |
+
+Testing needs to be performed to cover all branches, paths and conditions.
+
+```
+@Test
+fun testEvaluateAccess_adultMember() {
+  assertThat(evaluateAccess(25, true)).isEqualTo("Access granted")
+}
+
+@Test
+fun testEvaluateAccess_adultNonMember() {
+  assertThat(evaluateAccess(30, false)).isEqualTo("Membership required")
+}
+
+@Test
+fun testEvaluateAccess_minorMember() {
+  assertThat(evaluateAccess(16, true)).isEqualTo("Access denied")
+}
+
+@Test
+fun testEvaluateAccess_minorNonMember() {
+  assertThat(evaluateAccess(15, false)).isEqualTo("Access denied")
+}
+```
+
+Testing all branches and conditions ensures that your function can handle all possible scenarios, making it more reliable and easier to maintain.
+
+## 6. Exception and Error Handling
+
+**Exception Handling:**
+
+Exceptions are unexpected conditions or events that disrupt the normal flow of a program. They handle situations not part of the regular program flow, like invalid user input or file operation issues.
+
+Examples:
+- `NullPointerException` (accessing a null reference)
+- `IllegalArgumentException` (invalid function argument)
+- `IOException` (I/O operation failure)
+
+This allows to catch and handle issues, allowing the program to continue running or fail in a controlled manner.
+
+For instance, to handle division by zero error.
+
+```
+fun divide(a: Int, b: Int): Int {
+    if (b == 0) throw ArithmeticException("Division by zero")
+    return a / b
+}
+```
+
+**Error Handling:**
```

Comment:
Drop this part and mention how to handle domain errors, instead. Runtime errors (as in java.lang.Error) are outside the scope of usual testing since their boundary conditions are often either impossible to deterministically set up, or impractical (such as for OOM).

By 'errors' in my earlier comment I meant domain errors. For example: if a function can fail, but recover, it may print an error message and/or return a failure status. This is a failure in the sense that something went wrong, but it's not _exceptional_ in that execution needs to cease because the application is now in an irrecoverable state. It's really important to test both of these types of failure modes since they are both designed into implementations.

Reply:
Understood it based on the parsing error examples mentioned in the discussion and used those exact samples to present the sections. - https://github.com/oppia/oppia-android/blob/b1a5bd76878cfe21ce66ca01df984d0d904c4337/wiki/Writing-tests-with-good-behavioural-coverage.md#6-exception-and-error-handling